### PR TITLE
[MERGE] dev-apa into master

### PR DESCRIPTION
**Pull request by AntonSkrub**

### CODE REVIEW

**Review by valentin-kaiser — APPROVED** [review]

**Comment by valentin-kaiser** [comment]
Fix linter issues before merging

File: pkg/goGit/goGit.go [hunk 1]
```diff
@@ -11,14 +12,24 @@ import (
 	logr "github.com/sirupsen/logrus"
 )
 
-func UpdateLocalCopies(names []string, config *config.Config) {
+func UpdateLocalCopies(names []string, config *config.Config, user *config.User) {
 	for i := 0; i < len(names); i++ {
+		if i >= 10 {
+			logr.Info("Maximum number of repositories reached, exiting...")
+			break
+		}
```
Comment (valentin-kaiser) [inline pkg/goGit/goGit.go]: Remove this before merging into main

File: pkg/gitapi/gitapi.go [hunk 2]
```diff
@@ -12,7 +12,14 @@ import (
...
 
 func GetList(config *config.Config) []string {
-	req, err := http.NewRequest(http.MethodGet, "https://api.github.com/orgs/"+config.OrgaName+"repos", nil)
+	option := ""
+	if config.OrgaRepoType != "" {
+		option = "type=" + config.OrgaRepoType
+		logr.Infof("[API] Using type option: %v", option)
+	}
+
+	url := "https://api.github.com/orgs/" + config.OrgaName + "repos?" + option
```
Comment (valentin-kaiser) [inline pkg/gitapi/gitapi.go]: Use `url.JoinPath` and other functions from the package to create the URL. This minimises the chance of errors caused by malformed URLs.

**Review by valentin-kaiser — CHANGES_REQUESTED** [review]

File: pkg/goGit/goGit.go (hunk 1 above)
Comment (AntonSkrub) [reply pkg/goGit/goGit.go]: Resolved in 249b469

File: pkg/gitapi/gitapi.go (hunk 2 above)
Comment (AntonSkrub) [reply pkg/gitapi/gitapi.go]: Resolved in 249b469

File: pkg/goGit/goGit.go [hunk 3]
```diff
@@ -11,14 +13,18 @@ import (
...
 		if err != nil {
 			if err == git.ErrRepositoryNotExists {
 				logr.Errorf("[Git] Couldn't find a local copy of Repository %v", names[i])
-				Clone(names[i], config)
+
+				if user != nil {
+					Clone(names[i], config, user)
+				} else {
+					Clone(names[i], config, nil)
+				}
```
Comment (valentin-kaiser) [inline pkg/goGit/goGit.go]: Unnecessary If statement the third parameter 'User' is passed anyway in the called function.

File: pkg/goGit/goGit.go [hunk 4]
```diff
@@ -59,21 +64,23 @@ func UpdateLocalCopies(names []string, config *config.Config) {
...
+	if err != nil {
+		logr.Errorf("[GoGit] failed creating the url: %v\n", err)
+		return
+	}
 
+	auth := buildAuth(config, user)
 	// Clone the repository to the given directory
-	logr.Infof("[GoGit] Cloning the %v repository to %v", url+name+".git", config.OutputPath+name)
+	logr.Infof("[GoGit] Cloning the %v repository to %v", url, config.OutputPath+name)
 	r, err := git.PlainClone(config.OutputPath+name, false, &git.CloneOptions{
```
Comment (valentin-kaiser) [inline pkg/goGit/goGit.go]: Use a filepath build with `filepath.Join` instead of appending strings

File: pkg/goGit/goGit.go [hunk 5]
```diff
@@ -59,21 +64,23 @@ func UpdateLocalCopies(names []string, config *config.Config) {
 	}
 }
 
-func Clone(name string, config *config.Config) {
-	url := "https://github.com/" + config.OrgaName
+func Clone(name string, config *config.Config, user *config.User) {
+	url, err := url.JoinPath("https://github.com", name+".git")
```
Comment (valentin-kaiser) [inline pkg/goGit/goGit.go:64]: Use `url.JoinPath`

**Comment by valentin-kaiser** [comment]
Suggested edit:
```diff
diff --git a/pkg/goGit/goGit.go b/pkg/goGit/goGit.go
index c2d2335..afd30e2 100644
--- a/pkg/goGit/goGit.go
+++ b/pkg/goGit/goGit.go
@@ -94,17 +94,16 @@ func Clone(name string, config *config.Config, user *config.User) {
 }
 
 func buildAuth(config *config.Config, user *config.User) *http.BasicAuth {
-	var auth *http.BasicAuth
+	auth := &http.BasicAuth{
+		Username: config.OrgaName,
+		Password: config.OrgaToken,
+	}
+
 	if user != nil {
 		auth = &http.BasicAuth{
 			Username: user.Name,
 			Password: user.Token,
 		}
-	} else {
-		auth = &http.BasicAuth{
-			Username: config.OrgaName,
-			Password: config.OrgaToken,
-		}
 	}
 	return auth
 }

```

File: cmd/main.go [hunk 6]
```diff
@@ -14,18 +14,27 @@ import (
...
+		for _, user := range config.Users {
+			logr.Printf("[API] Found user: %v", user)
+			userRepoNames := gitapi.GetRepoList(config, &user)
+			logr.Info("Found ", len(userRepoNames), " repositories on the user account of ", user.Name)
+			gogit.UpdateLocalCopies(userRepoNames, config, &user)
+		}
+	}
 
 	UpdateInterval := cron.New()
 	UpdateInterval.AddFunc("*/3 * * * *", func() {
```
Comment (valentin-kaiser) [inline cmd/main.go]: The update interval should be configurable

**Review by valentin-kaiser — CHANGES_REQUESTED** [review]

**Comment by valentin-kaiser** [comment]
Suggested edit:
```diff
diff --git a/pkg/gitapi/gitapi.go b/pkg/gitapi/gitapi.go
index 8851d0a..0b2bc04 100644
--- a/pkg/gitapi/gitapi.go
+++ b/pkg/gitapi/gitapi.go
@@ -12,7 +12,7 @@ import (
 	logr "github.com/sirupsen/logrus"
 )
 
-type Repo struct {
+type Repository struct {
 	ID       int    `json:"id"`
 	Name     string `json:"name"`
 	FullName string `json:"full_name"`
@@ -20,7 +20,7 @@ type Repo struct {
 	Owner    string `json:"owner.login"`
 }
 
-func GetRepoList(config *config.Config, user *config.User) []string {
+func GetRepoList(config *config.Config, user *config.User) []Repository {
 	token, option, reqUrl := "", "", ""
 	var err error
 	if user != nil {
@@ -62,18 +62,12 @@ func GetRepoList(config *config.Config, user *config.User) []string {
 	}
 
 	// Unmarshal the json response to get the repository names
-	var repos []Repo
+	var repos []Repository
 	err = json.Unmarshal(body, &repos)
 	if err != nil {
 		logr.Errorf("[API] failed unmarshalling the json: %v\n", err)
 	}
 
-	i := 0
-	var repoNames []string
-	for _, repo := range repos {
-		i++
-		repoNames = append(repoNames, repo.FullName)
-	}
-	logr.Printf("[API] Found %v Repositories!", i)
-	return repoNames
+	logr.Printf("[API] Found %v Repositories!", len(repos))
+	return repos
 }
diff --git a/pkg/goGit/goGit.go b/pkg/goGit/goGit.go
index c2d2335..c4631e7 100644
--- a/pkg/goGit/goGit.go
+++ b/pkg/goGit/goGit.go
@@ -6,6 +6,7 @@ import (
 	"path/filepath"
 
 	"github.com/AntonSkrub/GoGit-Integration/pkg/config"
+	"github.com/AntonSkrub/GoGit-Integration/pkg/gitapi"
 
 	"github.com/go-git/go-git/v5"
 	"github.com/go-git/go-git/v5/plumbing/transport/http"
@@ -13,17 +14,17 @@ import (
 	logr "github.com/sirupsen/logrus"
 )
 
-func UpdateLocalCopies(names []string, config *config.Config, user *config.User) {
-	for i := 0; i < len(names); i++ {
-		r, err := git.PlainOpen(filepath.Join(config.OutputPath, names[i]))
+func UpdateLocalCopies(repos []gitapi.Repository, config *config.Config, user *config.User) {
+	for _, repo := range repos {
+		r, err := git.PlainOpen(filepath.Join(config.OutputPath, repo.FullName))
 		if err != nil {
 			if err == git.ErrRepositoryNotExists {
-				logr.Errorf("[Git] Couldn't find a local copy of Repository %v", names[i])
+				logr.Errorf("[Git] Couldn't find a local copy of Repository %v", repo.FullName)
 
 				if user != nil {
-					Clone(names[i], config, user)
+					Clone(repo.FullName, config, user)
 				} else {
-					Clone(names[i], config, nil)
+					Clone(repo.FullName, config, nil)
 				}
 			} else {
 				logr.Errorf("[Git] failed opening the repository: %v\n", err)
@@ -40,7 +41,7 @@ func UpdateLocalCopies(names []string, config *config.Config, user *config.User)
 		// Pull the latest changes from the origin and merge into the current branch
 		auth := buildAuth(config, user)
 
-		logr.Infof("[Git] Pulling the latest changes from the origin of %v", names[i])
+		logr.Infof("[Git] Pulling the latest changes from the origin of %v", repo.FullName)
 		err = w.Pull(&git.PullOptions{
 			RemoteName:   "origin",
 			SingleBranch: false,
@@ -49,7 +50,7 @@ func UpdateLocalCopies(names []string, config *config.Config, user *config.User)
 		})
 		if err != nil {
 			if err == git.NoErrAlreadyUpToDate {
-				logr.Errorf("[Git] Repository %v already up to date", names[i])
+				logr.Errorf("[Git] Repository %v already up to date", repo.FullName)
 			} else {
 				logr.Errorf("[Git] failed pulling the repository: %v\n", err)
 				return
@@ -60,7 +61,7 @@ func UpdateLocalCopies(names []string, config *config.Config, user *config.User)
 		if config.ListReferences || config.LogCommits {
 			AccessRepo(r, config)
 		}
-		logr.Infof("[Git] Finished updating the %v repository", names[i])
+		logr.Infof("[Git] Finished updating the %v repository", repo.FullName)
 	}
 }
 

```

File: pkg/goGit/goGit.go (hunk 4 above)
Comment (AntonSkrub) [reply pkg/goGit/goGit.go]: fixed in commit 0a1f9ae

File: pkg/goGit/goGit.go (hunk 3 above)
Comment (AntonSkrub) [reply pkg/goGit/goGit.go]: fixed in commit 0a1f9ae

File: pkg/goGit/goGit.go (hunk 5 above)
Comment (AntonSkrub) [reply pkg/goGit/goGit.go:64]: fixed in commit 0a1f9ae

File: cmd/main.go (hunk 6 above)
Comment (AntonSkrub) [reply cmd/main.go]: fixed in commit ddf71a5

**Review by valentin-kaiser — APPROVED** [review]

**Pull request merged**